crates/core/tedge_api/src/messages.rs

```diff
@@ -457,12 +460,62 @@ impl From<SoftwareError> for Option<SoftwareModuleItem> {
 /// Command to restart a device
 pub type RestartCommand = Command<RestartCommandPayload>;
 
+impl RestartCommand {
+    /// Create a restart command to be executed now (i.e. its status is scheduled)
+    /// On success the triggering command will be resumed in the `on_success` state.
+    /// On error the triggering command will be resumed in the `on_error` state.
+    pub fn with_context(
+        target: EntityTopicId,
+        cmd_id: String,
+        command: GenericCommandState,
+        on_success: StateName,
+        on_error: StateName,
+    ) -> Self {
+        let payload = RestartCommandPayload {
+            status: CommandStatus::Scheduled,
+            context: Some(RestartContext {
+                command,
+                on_success,
+                on_error,
+            }),
+        };
+        Command {
+            target,
+            cmd_id,
+            payload,
+        }
+    }
+
+    pub fn resume_context(&self) -> Option<GenericCommandState> {
+        self.payload
+            .context
+            .as_ref()
+            .and_then(|context| match self.status() {
+                CommandStatus::Init | CommandStatus::Scheduled | CommandStatus::Executing => None,
+                CommandStatus::Successful => Some(context.resume(None)),
+                CommandStatus::Failed { reason } => Some(context.resume(Some(reason))),
+            })
+    }
+}
+
 /// Command to restart a device
 #[derive(Debug, Clone, Default, Deserialize, Serialize, Eq, PartialEq)]
 #[serde(rename_all = "camelCase")]
 pub struct RestartCommandPayload {
     #[serde(flatten)]
     pub status: CommandStatus,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    context: Option<RestartContext>,
+}
+
+impl RestartCommandPayload {
+    pub fn new(status: CommandStatus) -> Self {
```

Comment:
```suggestion
    #[cfg(test)]
    pub fn new(status: CommandStatus) -> Self {
```
As this only gonna be used by the tests.